### PR TITLE
[Debug] Detect internal and deprecated methods

### PR DESCRIPTION
**Pull request by GuilhemN**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | no
| New feature?  | yes 
| BC breaks?    | no
| Deprecations? | no 
| Tests pass?   | yes
| Fixed tickets | 
| License       | MIT
| Doc PR        | 

I just realized we do not detect `@deprecated` methods like we do for `@final` methods, so here's a fix.

I reorganized the file to avoid code duplication... so the diff is kind of impressive but the behavior is the same.

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
rebase unlocked - the same should also apply to internal methods, isn't it?

**Comment by GuilhemN** [comment]
Yes, I'm on it.

**Comment by GuilhemN** [comment]
rebased

**Comment by GuilhemN** [comment]
fabbot failure is a false-positive. Travis and appveyor failures are unrelated.

File: src/Symfony/Component/Debug/DebugClassLoader.php [hunk 1]
```diff
@@ -361,4 +364,30 @@ public function loadClass($class)
...
+
+    /**
+     * `class_implements` includes interfaces from the parents so we have to
+     * manually exclude them.
+     *
+     * @param string $class
+     *
+     * @return string[]
+     */
+    private function getOwnInterfaces($class)
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Debug/DebugClassLoader.php]:
```php
        $ownInterfaces = class_implements($class, false);

        if ($parent = get_parent_class($class)) {
            foreach (class_implements($parent, false) as $interface) {
                unset($ownInterfaces[$interface]);
            }
        }

        foreach ($ownInterfaces as $interface) {
            foreach (class_implements($interface) as $interface) {
                unset($ownInterfaces[$interface]);
            }
        }

        return $ownInterfaces;
```

File: src/Symfony/Component/Debug/DebugClassLoader.php [hunk 2]
```diff
@@ -228,40 +183,88 @@ public function loadClass($class)
...
+                        list($methodShortName, $message) = self::$internalMethods[$use][$method->name];
+                        @trigger_error(sprintf('The "%s" method is considered internal%s. It may change without further notice. You should not use it from "%s".', $methodShortName, $message, $name), E_USER_DEPRECATED);
                     }
                 }
 
-                foreach ($deprecatedInterfaces as $interface) {
-                    if (!isset($parentInterfaces[$interface])) {
-                        @trigger_error(sprintf('The "%s" %s "%s" that is deprecated %s', $name, $refl->isInterface() ? 'interface extends' : 'class implements', $interface, self::$deprecated[$interface]), E_USER_DEPRECATED);
+                // Detect method annotations
+                $doc = $method->getDocComment();
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Debug/DebugClassLoader.php]: could be merged with previous line

File: src/Symfony/Component/Debug/DebugClassLoader.php [hunk 3]
```diff
@@ -361,4 +364,30 @@ public function loadClass($class)
             return true;
         }
     }
+
+    /**
+     * `class_implements` includes interfaces from the parents so we have to
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Debug/DebugClassLoader.php]: I think you can configure your IDE to auto-break lines at 120, not 80 :)

File: src/Symfony/Component/Debug/DebugClassLoader.php [hunk 4]
```diff
@@ -228,40 +183,88 @@ public function loadClass($class)
...
+                }
+            }
+
+            if ($parent = get_parent_class($class)) {
+                if (isset(self::$final[$parent])) {
+                    @trigger_error(sprintf('The "%s" class is considered final%s. It may change without further notice as of its next major version. You should not extend it from "%s".', $parent, self::$final[$parent], $name), E_USER_DEPRECATED);
+                }
+            }
+
+            $traits = class_uses($name, false);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Debug/DebugClassLoader.php]: `$traits = array($parent) + class_uses($name, false);`
Then below `foreach ($interfaces + $traits` to remove the array_merge() (or directly $traits, maybe renamed btw)

**Review by nicolas-grekas — APPROVED** [review]
with minor comments

File: src/Symfony/Component/Debug/DebugClassLoader.php (hunk 3 above)
Comment (GuilhemN) [inline src/Symfony/Component/Debug/DebugClassLoader.php]: Done :)

File: src/Symfony/Component/Debug/DebugClassLoader.php (hunk 2 above)
Comment (GuilhemN) [inline src/Symfony/Component/Debug/DebugClassLoader.php]: done

File: src/Symfony/Component/Debug/DebugClassLoader.php (hunk 4 above)
Comment (GuilhemN) [inline src/Symfony/Component/Debug/DebugClassLoader.php]: `array($parent) + $interfaces;` could cause conflicts so I did `$parentAndTraits = ($parent ? array($parent => $parent) : array()) + class_uses($name, false);` instead.

**Comment by nicolas-grekas** [comment]
Thank you @GuilhemN.

**Comment by stof** [comment]
This cause issue when you still need to extend the old method to keep support for older versions (look at the way bundles were able to support both Symfony 2.7 and 2.8 to configure options in form types, by implementing both the old `setDefaultOptions` for 2.7 and the new `configureOptions` for newer versions).

**Pull request closed without merging**